File: proxyGraper.pl
#!/usr/bin/perl
#ProxyGraper Coded By AntiGov

#Requires Modules
use strict;
use warnings;
use WWW::Mechanize;
use WWW::UserAgent::Random;
use Term::ANSIColor qw(:constants);
use Geo::IP;
use threads;
use Thread::Queue;
use MIME::Base64;
use v5.10;

sub info {

    #Display Banner
    my $banner = <<EOF;
______                    _____                           
| ___ \                  |  __ \                          
| |_/ / __ _____  ___   _| |  \/_ __ __ _ _ __   ___ _ __ 
|  __/ '__/ _ \ \/ / | | | | __| '__/ _` | '_ \ / _ \ '__|
| |  | | | (_) >  <| |_| | |_\ \ | | (_| | |_) |  __/ |   
\_|  |_|  \___/_/\_\\__, |\____/_|  \__,_| .__/ \___|_|   
__/ |               | |              
|___/ Coded By G66k |_|              
EOF
    say( BOLD, GREEN, $banner );
}
info();

# Some Variables
my ( $ip, $port, @proxy, @result );
my $page = 1;
say( BOLD RED, "# Please Wait Graping Proxy.." );
SecSite();
while ( $page <= 10 ) {    #greping 10 Pages max 67
    my $user_agent = rand_ua("browsers");     #random UserAgent;
    my $mech       = WWW::Mechanize->new();
    $mech->agent($user_agent);
    $mech->ssl_opts( verify_hostname => 0 );
    $mech->timeout(15);
    my $url = "http://www.proxylisty.com/ip-proxylist-$page";
    $mech->get($url);
    
    my $res = $mech->content;
    my @b   = parse($res);
    push( @result, @b );

    if ( $page == 10 ) {                      #check if we reach 10 Pages
        say( BOLD WHITE, "# Proxy Found: " . "[" . scalar(@result) . "]" );
        say( BOLD BLUE,  "# Checking Working Proxy.." );
        my $q = Thread::Queue->new();         #MultiTreads Start Here
        $q->enqueue($_) for @result;
        my $thread_limit
            = 10;    #Defauls Treads 10 Increase it if you have Good Machine
        my @thr = map {
            threads->create(
                sub {
                    while ( defined( my $ip = $q->dequeue_nb() ) ) {
                        check($ip);
                    }
                }
            );
        } 1 .. $thread_limit;
        $_->join() for @thr;
    }
    $page++;
}

sub SecSite {

    my @result;
    my $page = 0;
    while ( $page <= 10 ) {
        my $user_agent = rand_ua("browsers");
        my $mech = WWW::Mechanize->new( autocheck => 1 );
        $mech->agent($user_agent);
        $mech->ssl_opts( verify_hostname => 0 );
        $mech->timeout(15);
        my $url = "https://proxy-list.org/english/index.php?p=$page";
        $mech->get($url);
      sleep(2);

        my $res = $mech->content;


        my @b = parse2($res);
        push( @result, @b );
        if ( $page == 10 ) {    #check if we reach 10 Pages
            say( BOLD WHITE,
                "# Proxy Found: " . "[" . scalar(@result) . "]" );
            say( BOLD BLUE, "# Checking Working Proxy.." );
            my $q = Thread::Queue->new();    #MultiTreads Start Here
            $q->enqueue($_) for @result;
            my $thread_limit
                = 10;  #Defauls Treads 10 Increase it if you have Good Machine
            my @thr = map {
                threads->create(
                    sub {
                        while ( defined( my $ip = $q->dequeue_nb() ) ) {
                            if ( $ip eq '' ) { next; }
                            check($ip);
                        }
                    }
                );
            } 1 .. $thread_limit;
            $_->join() for @thr;
        }

        $page++;
    }

}

#Parsing Html content from proxy:port
sub parse {
    my $a = $_[0];
    my @proxy;
     my %x = ( $a =~ /<td>(\d+\.\d+\.\d+\.\d+)<\/td>\s+<td><a\shref='\S+'\stitle='\S+\s\S+\s\S+\s\S+'>(\d+)<\/a><\/td>/g );

        while(my ($k,$v) = each %x ){
            my $url = ($k.":".$v);
            push(@proxy,$url);

        }
      
    return @proxy;
}

sub parse2 {

    my $a = $_[0];
    my @proxy;
     my @x = ( $a =~ /class="proxy"><script\stype="text\/javascript">Proxy\('(.+)'\)<\/script>/g );
       
        foreach (@x) {
            my $url = decode_base64($_);
            
        push( @proxy, $url );

    }
    return @proxy;
}

#Checking Working Proxy
sub check {
    my @ips = @_;
    my $url = 'http://icanhazip.com';
    for ( my $i = 0; $i <= $#ips; $i++ ) {
        chomp( $ips[$i] );
        my $ua = LWP::UserAgent->new;
        $ua->proxy( 'http', "http://" . $ips[$i] );
        $ua->default_header( "Host" => "icanhazip.com" );
        $ua->agent(
            "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/536.6 (KHTML, like Gecko) Chrome/20.0.1092.0 Safari/536.6"
        );
        $ua->default_header( "Accept" =>
                "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
        );
        $ua->default_header( "Accept-Language" => "en-US,en;q=0.5" );
        $ua->default_header( "Accept-Encoding" => "gzip, deflate" );
        $ua->default_header( "Connection"      => "keep-alive" );
        $ua->default_header( "Cache-Control"   => "max-age=0" );
        $ua->timeout(30);
        my $response = $ua->get($url);

        if ( $response->is_success ) {
            ipInfo( $ips[$i] );
        }
    }
}

#Lookup Proxy
sub ipInfo {
    no warnings "uninitialized";    #Disable some anonying errors
    my $ip     = $_[0];
    my @r      = split( ":", $ip );
    my $gi     = Geo::IP->open( "GeoIP/GeoLiteCity.dat", GEOIP_STANDARD );
    my $record = $gi->record_by_addr( $r[0] );
    my $ok     = (
        BOLD YELLOW,
        $ip . " : " . BOLD WHITE,
        $record->country_code3 . " : "
            . $record->country_name . " : "
            . $record->region_name
            . BOLD GREEN,
        " [Ok]"
    );
    say($ok);
    open( SUCCESSFILE, '>>success.txt' );
    print SUCCESSFILE $ok . "\n";
    close(SUCCESSFILE);
}

File: proxyGraper.php
<?php
#!/usr/bin/php
#   Author: AntiGov   
#   ProxyGraper: Grap and check Proxy
#   @MA

$proxies = array();

for ($i=0; $i < 384; $i+=64) {
    preg_match_all('/(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})<\/td><td>(\d{1,5})/', _curl("https://hidemy.name/en/proxy-list/?type=hs&start=".$i."#list"), $matches);
    array_push($proxies, array_map("map_Proxy", $matches[1] , $matches[2]));
}

foreach ($proxies as $proxy) {
    if(is_array($proxy)){
        foreach ($proxy as $value) {
             if(check($value)){
                $ip = preg_split("/:/", $value);
                echo $value." ". getCountry($ip[0]). " Status: OK\n";
             }else{
                echo $value . " => Status: Dead\n";

             }
        }
    }
}
function map_Proxy($ip, $port){
    return($ip.":".$port);
}



function _curl($url,$post="",$usecookie = false) {  
    $ch = curl_init();
    if($post) {
        curl_setopt($ch, CURLOPT_POST ,1);
        curl_setopt ($ch, CURLOPT_POSTFIELDS, $post);
    }
    curl_setopt($ch, CURLOPT_URL, $url); 
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
    curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/6.0 (Windows; U; Windows NT 5.1; en-US; rv:1.7.7) Gecko/20050414 Firefox/1.0.3"); 
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
    if ($usecookie) { 
    curl_setopt($ch, CURLOPT_COOKIEJAR, $usecookie); 
    curl_setopt($ch, CURLOPT_COOKIEFILE, $usecookie);    
    } 
    curl_setopt($ch, CURLOPT_RETURNTRANSFER,1); 
    $result=curl_exec ($ch); 
    curl_close ($ch); 
    return $result; 
}

function check($proxy) {  
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, 'https://icanhazip.com/'); 
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
    curl_setopt($ch, CURLOPT_USERAGENT, "Mozilla/6.0 (Windows; U; Windows NT 5.1; en-US; rv:1.7.7) Gecko/20050414 Firefox/1.0.3"); 
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);
    curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 0);
    curl_setopt($ch, CURLOPT_CONNECTTIMEOUT ,5); 
    curl_setopt($ch, CURLOPT_TIMEOUT, 5);
    curl_setopt($ch, CURLOPT_PROXY, 'http://'.$proxy);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER,1); 
    $result=curl_exec ($ch); 
    curl_close ($ch); 
    if(preg_match('/(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})/', $result)){
        return 1;
    }
    return 0;
}
function getCountry($ip){
    $url = 'http://ip-api.com/json/'.$ip;
    $info = json_decode(_curl($url));
       if($info->status == "success"){
          return "Country: $info->country City: $info->city RegionName: $info->regionName";
       }
}

?>
